File: Java/src/main/java/com/gildedrose/BackstageItem.java
package com.gildedrose;

public class BackstageItem extends Item {

    public BackstageItem(String name, int sellIn, int quality) {
        super(name, sellIn, quality);
    }

    @Override
    protected void updateQualityBeforeSellIn() {
        increaseQualityIfBelow50();

        if (sellIn < 11) {
            increaseQualityIfBelow50();
        }

        if (sellIn < 6) {
            increaseQualityIfBelow50();
        }
    }

    protected void updateSellIn() {
        decreaseSellIn();
    }

    protected void updateQualityAfterSellIn() {
        if (sellIn < 0) {
            dropQuality();
        }
    }
}

File: Kotlin/src/main/kotlin/com/gildedrose/BackstageItem.kt
package com.gildedrose

class BackstageItem(override val name: String, override var sellIn: Int, override var quality: Int) : Item(name, sellIn, quality) {

    override fun updateQuality() {
        increaseQuality()
        if (this.sellIn < 11) {
            increaseQuality()
        }

        if (this.sellIn < 6) {
            increaseQuality()
        }
    }

    override fun updateSellIn() {
        decreaseSellIn()
    }

    override fun updateQualityWhenExpired() {
        dropQuality()
    }
}